keps/sig-api-machinery/2885-server-side-unknown-field-validation/README.md

```diff
@@ -0,0 +1,950 @@
+(i.e. we’re only focused on mismatched fields between the object in the request
+body and its schema).
+* Protobuf support. Theoretically, unknown protobuf fields could occur if clients
+on version X send a request to the server of version less than X which does not recognize
+some of the fields (or a similar situation). We do not think it is worth
+supporting this use case initially and will error if clients attempt to validate
+schema server-side with protobuf data.
+
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+We propose using an opt-in API mechanism (such as content-type header or query
+param) to indicate to the server that it should fail when the kubernetes object
+in the request body supplied to POST, PUT, and PATCH requests contains
+extra/unknown fields.
+
+Clients such as kubectl will continue to use the `--validate=true` flag as
+before, but instead of triggering validation on the client-side, it will
+instruct the server to validate for unknown fields server-side.
+
+This change will be made in at least two steps, one where we introduce the
+server side validation and a second where we modify kubectl to use the
+server-side validation (and mark the existing client-side validation as
+deprecated).
+
+### Notes/Constraints/Caveats (Optional)
+
+#### Future Work
+After server side unknown field validation is implemented we can begin work to
+deprecate and remove client side validate and have kubectl use server side
+validation instead. A separate KEP will be published for that.
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+### Performance Considerations
+It is worth noting that checking for unknown fields whether via a json unmarshal
+that explicitly breaks if it encounters unknown fields (as in the case of
+create, update, or json patch) or extra logic around a merge step (as in
+strategic merge patch or apply) will be less performant than not doing so.
+[Initial
+benchmarks](https://github.com/kubernetes/kubernetes/pull/104433#issuecomment-901398507) estimate ~20% slower 25-30% more memory consumption.
```

Comment:
I don't understand why strict serializer is so expensive.  While 20-30% is acceptable for opt in, the actual checks should be amortizable (we already have to iterate every field before assignment).  Naively I would expect that strict serialization should be capable of being free (<3-5%).

This is a statement about the general algorithm rather than the specific tradeoffs we're making in short term implementation.

Reply:
Yea, that's a fair point.

The 20-30% number comes from initial benchmarks on only CREATE/UPDATE requests. The only change there is that we use the [strict option](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go#L273-L293) on the json deserializer (which sets `disallowUnknownFields` in jsoniter).

I agree that we can and should investigate why this is so much less performant and see if we can do better, but as you've alluded to, in the short term it made more sense just to use what's already available and aim for better performance after we've green-lit the proposal.

Also worth noting, for PATCH (specifically json-patch/merge-patch and SMP) where we are the ones writing the unknown field validation code, I agree that I don't think we will see much if any performance impacts (but I didn't want to write this because I haven't confirmed it yet).